### ISSUE TITLE
phylotaR -- submission

### ISSUE THREAD

**Issue by DomBennett** [ropensci/onboarding#187]
### Summary

-   What does this package do? (explain in 50 words or less):

*R implementation of the [PhyLoTa](http://phylota.net/) pipeline. It identifies orthologous sequences in GenBank using BLAST searches. It consists of three automated stages: taxise, download and cluster.*

-   Paste the full DESCRIPTION file inside a code block below:

```
Package: phylotaR
Type: Package
Title: Automated phylogenetic sequence cluster identification from GenBank
Version: 0.1
Date: 2017-11-17
Author: Hannes Hettling, Rutger Vos, Alexander Zika, Dominic J. Bennett, Alexandre Antonelli
Maintainer: D.J. Bennett <dominic.john.bennett@gmail.com>
Description: PhyLoTa identifies orthologous sequence clusters from data readily available in GenBank. This is an R implementation of this process.
License: GPL-2
URL: https://github.com/dombennett/phylotaR#readme
BugReports: https://github.com/dombennett/phylotaR/issues
SystemRequirements: BLAST+ (>2.7.1)
Depends:
    R (>= 3.3.0),
    methods,
    foreach
Imports:
    curl,
    doMC,
    doSNOW,
    igraph,
    CHNOSZ,
    rentrez,
    DBI,
    taxize,
    XML
Suggests:
    testthat,
    knitr,
    rmarkdown
RoxygenNote: 6.0.1
VignetteBuilder: knitr
```

-   URL for the package (the development repository, not a stylized html page): https://github.com/DomBennett/phylotaR

- Please indicate which category or categories from our [package fit policies](https://github.com/ropensci/onboarding/blob/master/policies.md#package-fit) this package falls under ***and why**(? (e.g., data retrieval, reproducibility. If you are unsure, we suggest you make a pre-submission inquiry.):

*data retrieval, it downloads sequences from genbank and identifies sequence clusters that may be of use for phylogenetic analysis*

-   Who is the target audience and what are scientific applications of this package?  

*Evolutionary biologists, phylogeneticists. Determining sequence orthology is the essential first step of any phylogenetic analysis. Usually a phylogeneticist may search GenBank for sequences of the same gene, download and align. Due to naming problems, however, potential relevant sequences may be missed or sequences that have been mislabelled are downloaded. PhyLoTa gets around this by performing all-v-all BLAST searches and does not rely on gene names. PhyLoTa is often the first step of any large-scale automated phylogenetic pipeline (e.g. [SUPERSMART](http://www.supersmart-project.org/)). Potentially, this R package could be central to many future phylogenetic analyses.*

-   Are there other R packages that accomplish the same thing? If so, how does
yours differ or meet [our criteria for best-in-category](https://github.com/ropensci/onboarding/blob/master/policies.md#overlap)?

*The only current approach for identifying PhyLoTa clusters is the PhyLoTa browser. The data underlying its current release, however, is five years old. In those five years there have been [over 40 million new sequences](https://www.ncbi.nlm.nih.gov/genbank/statistics/) added to GenBank. An alternative is needed to make use of the latest data. Additionally, phylotaR would allow a user to specify their own parameters when running the pipeline.*

-   If you made a pre-submission enquiry, please paste the link to the corresponding issue, forum post, or other discussion, or @tag the editor you contacted.

### Requirements

Confirm each of the following by checking the box.  This package:

- [x] does not violate the Terms of Service of any service it interacts with. 
- [x] has a CRAN and OSI accepted license.
- [x] contains a README with instructions for installing the development version. 
- [x] includes documentation with examples for all functions.
- [x] contains a vignette with examples of its essential functions and uses.
- [x] has a test suite.
- [x] has continuous integration, including reporting of test coverage, using services such as Travis CI, Coveralls and/or CodeCov.
- [x] I agree to abide by [ROpenSci's Code of Conduct](https://github.com/ropensci/onboarding/blob/master/policies.md#code-of-conduct) during the review process and in maintaining my package should it be accepted.

#### Publication options

- [x] Do you intend for this package to go on CRAN?  
- [ ] Do you wish to automatically submit to the [Journal of Open Source Software](http://joss.theoj.org/)? If so:
    - [ ] The package has an **obvious research application** according to [JOSS's definition](http://joss.theoj.org/about#submission_requirements).
    - [ ] The package contains a `paper.md` matching [JOSS's requirements](http://joss.theoj.org/about#paper_structure) with a high-level description in the package root or in `inst/`.
    - [ ] The package is deposited in a long-term repository with the DOI: 
    - (*Do not submit your package separately to JOSS*)
- [ ] Do you wish to submit an Applications Article about your package to [Methods in Ecology and Evolution](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/)? If so:
	- [ ] The package is novel and will be of interest to the broad readership of the journal.
	- [ ] The manuscript describing the package is no longer than 3000 words.
	- [ ] You intend to archive the code for the package in a long-term repository which meets the requirements of the journal.
	- (*Please do not submit your package separately to Methods in Ecology and Evolution*)

### Detail

- [x] Does `R CMD check` (or `devtools::check()`) succeed?  Paste and describe any errors or warnings:

*No errors, but plenty of warnings to do with documentation.*

- [ ] Does the package conform to [rOpenSci packaging guidelines](https://github.com/ropensci/onboarding/blob/master/packaging_guide.md)? Please describe any exceptions:

*Not just yet. Outstanding: conversion to snake_case, generate README from rmd, create website with pkgdown, NEWS required, authors@R syntax, depends to imports and cat to message (although is that necessary given the logging methods?)*

- If this is a resubmission following rejection, please explain the change in circumstances:

- If possible, please provide recommendations of reviewers - those with experience with similar packages and/or likely users of your package - and their GitHub user names:

## Pre-submission Note

*Hi!
I've produced this pre-submission issue to determine whether phylotaR would be something you'd like added to your roster? On our side, we would be very keen as we think future phylogenetic analyses will depend on making software open-source, modular and collaborative. I did feel, however, that it may not quite fit into the ropensci family, because it's a pipeline and it depends on software external to R (BLAST). Please be aware that phylotaR is still only being developed and a lot of testing still needs to be performed. If you are interested, I'd be happy to update the naming style (I prefer camelcase), improve the documentation and whatever else is needed to meet your criteria.
Thanks,
Dom*

**Comment by sckott** [comment]
thanks for your submission @DomBennett we've been discussing and will get back to you asap

**Comment by sckott** [comment]
### Editor checks:

- [x] **Fit**: The package meets criteria for [fit](policies.md#fit) and [overlap](policies.md#fit)
- [x] **Automated tests:** Package has a testing suite and is tested via Travis-CI or another CI service.
- [x] **License:** The package has a CRAN or OSI accepted license
- [x] **Repository:** The repository link resolves correctly
- [ ] **Archive** (JOSS only, may be post-review): The repository DOI resolves correctly
- [ ] **Version** (JOSS only, may be post-review): Does the release version given match the GitHub release (v1.0.0)?

---

#### Editor comments

Thanks for your submission @DomBennett   !! 

- You didn't check the top level MEE box - i've checked that box. You didn't check the box for `The manuscript describing the package is no longer than 3000 words` - will manuscript length be a problem?
- You didn't check the box for `has a CRAN and OSI accepted license.` I think you do, can you check that box?
- You didn't check the box for `does not violate the Terms of Service of any service it interacts with.` Is that going to be a problem? Can you check on that?
- I see you checked the submit to MEE box. Note that we can not make any gaurantee about MEE considering your paper a good fit for them, or them accepting the paper
- Seems like `blast` on Homebrew (installed via `brew install blast`) is the same as at the NCBI link you gave at https://www.ncbi.nlm.nih.gov/books/NBK279671/  - Is that right? homebrew is for OSX platforms only
- It would indeed be ideal if the pkg didn't have to use system calls to external tools, but there are times when there's no other way.
- In terms of style, we prefer snake case but as long as you are consistent within the package that is THE most important thing (i.e, don't use a combination of camelCase, TitleCase and snake_case). 
- You did not check the box for `Does the package conform to rOpenSci packaging guidelines?` Can you do that and describe any exceptions
- You can put a ropensci review badge in your README

`[![](https://badges.ropensci.org/187_status.svg)](https://github.com/ropensci/onboarding/issues/187)`

- Goodpractice output for you and reviewers to consider:

```r
── GP phylotaR ────────────

It is good practice to

  ✖ write unit tests for all functions, and all package code
    in general. 75% of code lines are covered by test cases.

    R/base-wrappers.R:7:NA
    R/blast-tools.R:10:NA
    R/blast-tools.R:16:NA
    R/blast-tools.R:28:NA
    R/blast-tools.R:34:NA
    ... and 325 more lines

  ✖ not use "Depends" in DESCRIPTION, as it can cause name
    clashes, and poor interaction with other packages. Use "Imports"
    instead.
  ✖ omit "Date" in DESCRIPTION. It is not required and it
    gets invalid quite often. A build date will be added to the package
    when you perform `R CMD build` on it.
  ✖ use '<-' for assignment instead of '='. '<-' is the
    standard, and R users and developers are used it and it is easier
    to read your code for them if you use '<-'.

    R/ncbi-tools.R:156:9

  ✖ avoid long code lines, it is bad for readability. Also,
    many people prefer editor windows that are about 80 characters
    wide. Try make your lines shorter than 80 characters

    R/blast-tools.R:59:1
    R/taxise-tools.R:361:1

  ✖ omit trailing semicolons from code lines. They are not
    needed and most R coding standards forbid them

    R/ncbi-tools.R:147:55

  ✖ avoid sapply(), it is not type safe. It might return a
    vector, or a list, depending on the input data. Consider using
    vapply() instead.

    R/blast-tools.R:32:13
    R/blast-tools.R:33:11
    R/cache-tools.R:253:13
    R/cache-tools.R:256:13
    R/cache-tools.R:317:12
    ... and 32 more lines

  ✖ avoid 1:length(...), 1:nrow(...), 1:ncol(...),
    1:NROW(...) and 1:NCOL(...) expressions. They are error prone and
    result 1:0 if the expression on the right hand side is zero. Use
    seq_len() or seq_along() instead.

    R/cluster-tools.R:345:12
    R/ncbi-tools.R:162:20
    R/pipeline.R:208:12
    tests/testthat/test-cluster-tools.R:123:17
    tests/testthat/test-download-tools.R:81:17
    ... and 2 more lines

  ✖ fix this R CMD check WARNING: '::' or ':::' imports not
    declared from: ‘ggplot2’ ‘plyr’ Namespaces in Imports field not
    imported from: ‘DBI’ ‘taxize’ All declared Imports should be used.
    Packages in Depends field not imported from: ‘foreach’ ‘methods’
    These packages need to be imported from (in the NAMESPACE file) for
    when this namespace is loaded but not attached. Missing or
    unexported object: ‘doSNOW::registerDoMC’
  ✖ fix this R CMD check NOTE: blstN: no visible binding for
    global variable ‘cmd’ blstN: no visible global function definition
    for ‘read.table’ byLineage : .calc : <anonymous>: no visible
    binding for global variable ‘value’ fltr: no visible global
    function definition for ‘head’ fltr: no visible global function
    definition for ‘tail’ genClstrsObj: no visible global function
    definition for ‘new’ getBestClstrs: no visible global function
    definition for ‘nTaxa’ getBestClstrs: no visible global function
    definition for ‘seqLength’ getClMdLn : <anonymous>: no visible
    global function definition for ‘median’ getMngblIds: no visible
    global function definition for ‘head’ getMngblIds: no visible
    global function definition for ‘tail’ mkBlstDB: no visible binding
    for global variable ‘cmd’ nDscndnts: no visible global function
    definition for ‘%dopar%’ nDscndnts: no visible global function
    definition for ‘foreach’ nDscndnts: no visible binding for global
    variable ‘i’ plotTable: no visible binding for global variable
    ‘clstrnm’ plotTable: no visible binding for global variable
    ‘txidnm’ plotTable: no visible binding for global variable ‘value’
    runStgs: no visible global function definition for ‘is’ safeSrch:
    no visible global function definition for ‘is’ setUp: no visible
    global function definition for ‘packageVersion’ writeClstr: no
    visible global function definition for ‘write.table’ writeSqs: no
    visible binding for global variable ‘clstrs_obj’ writeTax: no
    visible global function definition for ‘write.table’
    summary,ClstrsObj: no visible binding for global variable ‘x’
    Undefined global functions or variables: %dopar% clstrnm clstrs_obj
    cmd foreach head i is median nTaxa new packageVersion read.table
    seqLength tail txidnm value write.table x Consider adding
    importFrom("methods", "is", "new") importFrom("stats", "median")
    importFrom("utils", "head", "packageVersion", "read.table", "tail",
    "write.table") to your NAMESPACE file (and ensure that your
    DESCRIPTION Imports field contains 'methods').
  ✖ fix this R CMD check WARNING: Undocumented data sets:
    ‘blst_rs’ ‘phylt_nds’ ‘sqs’ ‘tdobj’ All user-level objects in a
    package should have documentation entries. See chapter ‘Writing R
    documentation files’ in the ‘Writing R Extensions’ manual.
  ✖ fix this R CMD check NOTE: Warning: no function found
    corresponding to methods exports from ‘phylotaR’ for: ‘show’ A
    namespace must be able to be loaded with just the base namespace
    loaded: otherwise if the namespace gets loaded by a saved object,
    the session will be unable to start. Probably some imports need to
    be declared in the NAMESPACE file.
```

You don't have to address these now. 

Seeking reviewers now 🕐 

---

Reviewers:  @xx @xx 
Due date: xxxx

**Comment by rvosa** [comment]
@sckott - what would you recommend for a journal to publish a short application note about this package?

**Comment by sckott** [comment]
@rvosa possibly: we have a sort of pairing with JOSS and MEE. Also maybe: The R Journal, F1000Research

**Comment by rvosa** [comment]
Thanks! Sorry, what's JOSS?

On Wed, Jan 31, 2018 at 8:33 PM, Scott Chamberlain <notifications@github.com

**Comment by sckott** [comment]
http://joss.theoj.org/

**Comment by sckott** [comment]
@DomBennett I had a number of questions for you above in my comments. Please respond

**Comment by DomBennett** [comment]
@sckott Sorry, I thought I could wait for the reviewers and respond to everything wholesale.

**You didn't check the top level MEE box - i've checked that box. You didn't check the box for The manuscript describing the package is no longer than 3000 words - will manuscript length be a problem?**

We hadn't decided at that point which journal to send it to. We're not likely to submit to MEE now. All MEE boxes have been unticked.

**You didn't check the box for has a CRAN and OSI accepted license. I think you do, can you check that box?**

It does now. I added an MIT license.

**You didn't check the box for does not violate the Terms of Service of any service it interacts with. Is that going to be a problem? Can you check on that?**

I've read the terms now and it won't. Box ticked.

**I see you checked the submit to MEE box. Note that we can not make any gaurantee about MEE considering your paper a good fit for them, or them accepting the paper.**

It'd be nice if you did guarantee! But we're not going with MEE....

**Seems like blast on Homebrew (installed via brew install blast) is the same as at the NCBI link you gave at https://www.ncbi.nlm.nih.gov/books/NBK279671/ - Is that right? homebrew is for OSX platforms only**

It's just the link to the general description of how to install BLAST tools across all systems. Not specific to OSX.

**It would indeed be ideal if the pkg didn't have to use system calls to external tools, but there are times when there's no other way.**

Yeah it sucks big time. We're currently working on an approach to avoid using standalone BLAST, in which case a user would not need to install it. The BLAST tools are written in C++, so in theory it should be possible to package the code into the R package. I'm not too au fait on how to do this though. I've written external C scripts into R packages before, but never have I taken someone else's C code entirely. There may be problems with modifying the C++ code to interact with the R. Do you have any ideas on workarounds? Otherwise we could simply give the user the option to send their BLAST jobs to online servers (e.g. NCBI or EBI).

**In terms of style, we prefer snake case but as long as you are consistent within the package that is THE most important thing (i.e, don't use a combination of camelCase, TitleCase and snake_case).**

Yeah I know..... sigh..... Coding style can be real bugbears. I try and use a sort of PEP8 style with a mix of all three (variables snake_case, functions camelCase and classes TitleCase). I can go through and make it all snake_case. I'll probably rechristen some elements where I think the names don't quite work.

**You did not check the box for Does the package conform to rOpenSci packaging guidelines? Can you do that and describe any exceptions**

Not just yet. See above. Also I need to go through your more specific recommendations above as well. Do I need to make these changes before the reviewers look over the package? It's just the pipeline is still being tested and developed and it may be a while before these changes can be made.

**You can put a ropensci review badge in your README**

Dun'd.

**Comment by sckott** [comment]
> We hadn't decided at that point which journal to send it to. We're not likely to submit to MEE now. All MEE boxes have been unticked.

Okay, sounds good. 

> It does now. I added an MIT license.

thanks. 

> I've read the terms now and it won't. Box ticked.

thanks.

> We're currently working on an approach to avoid using standalone BLAST ...

Glad you're looking into some options. It's possible reviewers will have some ideas as well.  One idea: if you do shell out, you could use the `sys` package made by Jeroen on our team https://cran.rstudio.com/web/packages/sys/ 

> I need to go through your more specific recommendations above as well. Do I need to make these changes before the reviewers look over the package?

If you're talking about the `goodpractice` output, no, not now. That's just information for you and reviewers to consider.

**Comment by arendsee** [comment]
Hi @DomBennett, I'm working on reviewing `phylotaR`, but ran into a few little issues.

 1.  When I run the vignette, I get the following error:

```R
> clstr_id <- getBestClstrs(clstrs_obj, n=1)                        
Error in nTaxa(clstrs_obj) : could not find function "nTaxa"
```

 2. Fix the BLAST version handling

Currently the BLAST version requirement is hard coded to greater than 2.7.1
(`setup-tools.R:118`). I have 4 issues with this:

 * Hard-coding of values is generally bad.

 * BLAST version 2.7.1 is pretty restrictive. For example, the Ubuntu package
 manager provides BLAST version 2.2.28. Many researchers work in paranoid
 agencies that don't give us admin privileges, so installing the latest and
 greatest is often a pain. Anyway, unless your algorithm depends on new
 features, you should relax the version requirement.

 * The code for testing version is brittle. The version vector is parsed into a 3-tuple (e.g. 2.7.1 ==> `c(2,7,1)`), then tested as below

```R
  # setup-tools.R:118
  tst <- vrsn[1] >= 2 & vrsn[2] >= 7 & vrsn[3] >= 1
```

 However, this would lead to versions such as `2.8.0` being rejected. You can
 fix this by dropping the last comparison:

```R
  # setup-tools.R:118
  tst <- vrsn[1] >= 2 & vrsn[2] >= 7
```

Also, in DESCRIPTION, the  `SystemRequirements` specifies a BLAST version of `>2.7.1`, but your actual version checking function requires `>=2.7.1`.

 3. Fix the warnings and notes in `check()`

 When I run `devtools::check()` I get 5 warnings, and 3 notes. The issues all look pretty normal (easy to fix). I don't see anything too serious here. Below I have outlined the main issues and how to fix them (I might have missed one or two).

 * Remove the `warnings_are_errors: false` clause from your `.travis.yml`
 file. CRAN will generally not accept a package if any warnings are raised by
 CHECK. So you probably should not ignore warnings in your continuous
 integration.

 * DESCRIPTION says that your license is GPL-2, but the actual LICENSE file is
 MIT. You need to choose one of them. `devtools` makes license handling pretty
 easy. Just delete your current LICENSE file and the `License` entry in your
 DESCRIPTION file, and then call either `devtools::use_mit_license()` or
 `devtools::use_gpl3_license()`. This will create the license and update the
 DESCRIPTION accordingly.

 * You need to add `ggplot2` and `plyr` to your `Imports` list in DESCRIPTION

 * You need to add the statements:

```R
#' @importFrom methods is new
#' @importFrom stats median 
#' @importFrom utils heads packageVersion read.table tail write.table
```

 Somewhere in your documentation (I usually put these lines in the same place
 where I put package level documentation).

 * Describe each parameter in the documentation

Cheers!

**Comment by DomBennett** [comment]
Thanks for the detailed and early feedback. Dealing with your specific issues:

1.  nTaxa
The functions and classes associated with the clusters object are still a little open-ended at this stage, hence why their names may change. The equivalent to `nTaxa` is now `getClNTx` - get n taxa by cluster. We're still trying to determine the sorts of ways users will want to interrogate the clusters, and the best sort of language to use when naming functions.

I've updated the vignette. But the code is likely to change in the near future and will need changed again.

2. BLAST
OK. Thanks for the tips. Fixed and updated. Now using v 2.0 and up. How would you test the versioning without hard coding the values?

3. Check warnings

* travis warnings - removed from yaml
* licensing updated to MIT
* ggplot2 and plyr - added
* statements added to phylotaR.R
* parameters are being worked and should be updated soon!

**Comment by sckott** [comment]
Both reviewers now assigned:  @arendsee @naupaka
Due date: 2018-03-12

**Comment by DomBennett** [comment]
@arendsee @naupaka Thanks for taking time to review the package.

**Comment by arendsee** [comment]
> OK. Thanks for the tips. Fixed and updated. Now using v 2.0 and up. How would you test the versioning without hard coding the values?

Hmm, good point. @sckott Any thoughts on best practices for confirming the version of an outside tool (BLAST, in this case)?

I can tell you what I do, though I am not entirely happy with my approach. I simply don't check the version of my tools, instead I just print the version output into the log (e.g. system call to `blastp -version`).

If you want to check the version, you could write a dedicated function to extract a version string from a version statement (maybe `get_version` or something), and then write a dedicated version comparison function (maybe `cmp_version`).

I don't really have a good general solution. But in your case, I would recommend not checking the version. There are differences between the BLAST versions, but I don't _think_ any of the BLAST+ versions anyone will actually use (i.e. >=2.2) will change the results. Perhaps it would be worthwhile to test several versions of BLAST on Travis.

Thoughts? @sckott @naupaka?

**Comment by sckott** [comment]
I don't know off hand about best practices for checking versions of cli tools. Thoughts @jeroen @richfitz ?

> Perhaps it would be worthwhile to test several versions of BLAST on Travis.

this seems like a really good idea - and then any changes that affect certain BLAST versions and not others will become immediately apparent

**Comment by naupaka** [comment]
Insofar as I can recall, the big breaking changes happened during the switch from BLAST to BLAST+ that happened ~6 years ago. So as long as the version is newer than that it should be ok. Some of the other NCBI software like sra-tools need to have more recent versions than are currently available in some apt repositories, because of changes in NCBI's servers, but I haven't seen problems with BLAST itself. 

It seems (in my personal opinion) that trying to pull the source for BLAST into R is a bad idea, because then it's a ton of extra maintenance to make sure that stays up to date as things change on NCBI's side. Most folks who would use this package should probably be able to install a modern copy of ncbi-blast, so I would argue for spending effort on increasing the normal codebase and adding documentation, etc, instead of trying to pull BLAST in to remove the dependency.

**Comment by arendsee** [comment]
@naupaka I agree on pulling BLAST into R being a bad idea. Unless NCBI was interested in doing the work and maintaining the code.

As for the legacy BLAST versus BLAST+, the names of the executable have all changed. For example, `blastp` is not defined in legacy, they used `blastall` with some protein-vs-protein argument, I think. So if someone tried to use legacy BLAST the commands required simply would not be found.

**Comment by DomBennett** [comment]
OK. So... actionable points from this discussion:
- Don't test for BLAST version
  - Print version in log
  - But do ensure it's not legacy BLAST
- Write tests that run across different BLAST versions for travis
- Petition NCBI to support BLAST in R

**Comment by arendsee** [comment]
@DomBennett I probably got a bit beyond the scope of this review in my earlier comments. I'd say the only actionable point in regards to `phylotaR` is the first bullet.

Tests in Travis against multiple versions of BLAST would be cool, but as far as I am concerned, you don't need to do that now.

**Comment by sckott** [comment]
@arendsee @naupaka bump, your review is due in about one week

**Comment by arendsee** [comment]
@sckott Right! I'll get on it, thanks for the reminder

**Comment by naupaka** [comment]
Right-o. It's on the docket for this weekend.

**Comment by arendsee** [comment]
This is solid work and I recommend it be admitted to rOpenSci after a few
issues are addressed. The documentation looks good, program seems to work as
intended, the interface to the output objects is clear, and the results are
neatly cached and logged. This is a good example of how a clean R pipeline
should look.

My one high-level comment is that this package could benefit greatly by reusing
some existing R packages: `taxize` and `taxizedb` can handle NCBI taxonomy IDs;
bioconductor's `Biostrings` can be used for reading, writing, and translating
sequence; possibly `drake` could be used as a build engine.


# Package Review

Please check off boxes as applicable, and elaborate in comments below. Your
review is not limited to these topics, as described in the reviewer guide

 * [x] As the reviewer I confirm that there are no conflicts of interest for me
 to review this work (If you are unsure whether you are in conflict, please
 speak to your editor before starting your review).

## Documentation

The package includes all the following forms of documentation:

 * [x] A statement of need clearly stating problems the software is designed to
 solve and its target audience in README

 * [x] Installation instructions: for the development version of package and any
 non-standard dependencies in README

 * [x] Vignette(s) demonstrating major functionality that runs successfully
 locally

 * [x] Function Documentation: for all exported functions in R help

 * [ ] Examples for all exported functions in R Help that run successfully locally

**No examples given**

 * [ ] Community guidelines including contribution guidelines in the README or
 CONTRIBUTING, and DESCRIPTION with URL, BugReports and Maintainer (which may
 be autogenerated via Authors@R).

**Missing CONTRIBUTING**

## Functionality

 * [x] Installation: Installation succeeds as documented.

 * [x] Functionality: Any functional claims of the software been confirmed.

 * [x] Performance: Any performance claims of the software been confirmed.

There aren't really any performance claims made. However, the vignette ran
pretty quickly.

 * [ ] Automated tests: Unit tests cover essential functions of the package and
 a reasonable range of inputs and conditions.

Testing coverage is at 49%. I'd recommend spending some time nailing all the
functions down with direct comparisons to their expected results. It is also
worthwhile to check all the corner cases.

 * [x] All tests pass on the local machine.

 * [ ] Packaging guidelines: The package conforms to the rOpenSci packaging
 guidelines

 **- missing examples**

## Final approval (post-review)

 * [ ] The author has responded to my review and made changes to my satisfaction.
 I recommend approving this package.

Estimated hours spent reviewing: 6 hours

-------------------------------------------------------------------------------

## Review Comments

### Contributing

`rOpenSci` recommends you add a CONTRIBUTING file to your repo.

### DESCRIPTION

 * [ ] I would recommend a version formatted as `<major>.<minor>.<patch>` (e.g.
 `0.1.0`), [see here](http://r-pkgs.had.co.nz/description.html#version). This
 is by far the most common convention in the R world (and in the open source
 world in general).

 * [ ] The `Date` field should not be used

 * [ ] For the `URL` and `BugReports` entries, it would be better if these
 pointed to the repository that holds the source code (see `Repos issues`
 section below). 

 * [ ] The `License` slot says GPL-2 but your actual LICENSE is MIT, this needs
 to be resolved.

### Style

Skimming the code, everything looks tidy and well organized. Functions are
partitioned between files in an understandable way and every function is
documented (see `CHECK` for exceptions to this). I have just a few little
comments:

 * [ ] Your line wrapping is a little irregular, for example, in the description
 of the `run` function. Some function documentation is wrapped at 80 lines,
 some at 70, and some at 50 (e.g. `runTaxise`). This is not a big deal, but it
 is a little awkward to read.

 * [ ] There is an extra newline at the beginning of some, but not all, of your files.

### Documentation

The documentation is pretty solid. Just a few comments:

 * [ ] There are a few miss-spellings in the documentation (e.g.
 `cluster-tools.R:340`). You should run `devtools::spell_check()` and fix all
 mistakes (don't worry about the false positives).

 * [ ] Since R is dynamically typed, it is often useful to specify the expected
 type for each parameter and the return type in the documentation. The input
 types are covered reasonably well in the `@param` fields, but a lot of
 functions are missing the `@return` field. I strongly recommend adding
 a `return` field to every function, even if it returns nothing.

 * [ ] According to the rOpenSci guidelines, each function should have examples
 in the documentation. You might consider creating example data for your
 package, that can be loaded and used in the examples. For example, you could
 wrap all the data from your vignette example into a list, and then load it
 with `data(aotus)`. The `devtools::use_data` function makes this pretty easy
 to do. This approach circumvents the need to rerun a whole analysis in each
 example.

 * [ ] Some of the function and class names are a little cryptic. I would prefer
 longer, more descriptive and memorable names, for example `ClusterRecord`
 rather than `ClRcrd` (which I might interpret as `CladeRecord`, among other
 possibilities).

 * [ ] Snakecase is usually preferred in R for variable names. I'm fine with
 class names being camel case (that is how bioconductor does it).

### Repo issues

 The `DomBennett/phylotaR` repo is forked from the `AntonelliLab/phylotaR`
 repo. According to DESCRIPTION, issues should be logged at
 `AntonelliLab/phylotaR`. I don't think this is a good idea -- issues should
 not be separated from their source. There are at least two solutions:

 1. Allow issues to be posted on the `DomBennett/phylotaR` repo. By default you
    cannot post issues on a forked repo. However, this default can be
    [overridden](https://softwareengineering.stackexchange.com/questions/179468).
    
 2. Merge the fork back into `AntonelliLab/phylotaR`.

### README

Overall the README looks good. Here are a few suggestions:

  * [ ] Delete the comment, "Please note: this package is in its development and
  is not fully tested.". Before this goes into rOpenSci, you should be
  reasonably confident that the program works. There are always bugs, but we
  everyone expects this, so no special mention is needed.

  * [ ] Specify the minimum required BLAST+ version. 

  * [ ] In the `Pipeline` section, I might break the first paragraph into
  a numbered list of steps; then elaborate a *little* more. Specifically, you
  should say how you identify orthologous clusters with BLAST. Then maybe
  another sentence of explanation on `cluster^2`. Overall, the `Pipeline`
  section is great and I like your figure. 

  * [ ] The naming you use in the `Pipeline` section is a little confusing.
  I recommend using the same names you use in your code. Rather than saying
  `taxise`, `download`, `cluster`, and `cluster^2`; say `runClusters`,
  `runClusters2`, `runDownload`, and `runTaxise`. This is especially confusing
  for `taxise`, where I think you are referring to `runTaxise` which is
  performing a function that is very similar to what might be done in the
  `taxize` package (which mixed me up at first).

### Vignette

The introductory vignette is well written and clear. There is a reference to an
advanced vignette, but this vignette is not yet written. Here are my
recommendations:

  * [ ] In the introductory vignette, add a section discussing how long the
  program will take to run.

  * [ ] Either open a github issue or a TODO list entry for completion of the
  advanced vignette. As far as my review is concerned, I am OK with this
  submission lacking the vignette.

### Running the code

Here I will walk through the vignette and add suggestions in checkboxes.

```R
library(phylotaR)
wd <- 'aotus'
ncbi_dr <- '/usr/bin'
txid <- 9504
setUp(wd=wd, txid=txid, ncbi_dr=ncbi_dr, v=TRUE)
#R> -----------------------------------------------
#R> phylotaR: Implementation of PhyLoTa in R * [v0.1]
#R> -----------------------------------------------
#R> Checking for valid NCBI BLAST+ Tools ...
#R> Found: [/usr/bin/makeblastdb]
#R> Found: [/usr/bin/blastn]
#R> Setting up pipeline with the following parameters:
#R> . blstn      [/usr/bin/blastn]
#R> . btchsz     [300]
#R> . date       [2018-03-11]
#R> . mdlthrs    [3000]
#R> . mkblstdb   [/usr/bin/makeblastdb]
#R> . mncvrg     [51]
#R> . mnsql      [250]
#R> . mxevl      [1e-10]
#R> . mxnds      [1e+05]
#R> . mxrtry     [100]
#R> . mxsql      [2000]
#R> . mxsqs      [50000]
#R> . ncps       [1]
#R> . txid       [9504]
#R> . v          [TRUE]
#R> . wd         [aotus]
#R> -----------------------------------------------
```

This output summary is concise and informative and is recorded in a log file.

 * [ ] The log should include the NCBI BLAST version.

 * [ ] It also might be nice to include the `sessionInfo()` results in the log.
 Or you could write it into a separate file, if it clutters the log too much.

Next we run the program:

```R
run(wd=wd)
```

This prints results to the screen and the log. There are quite a few messages
like this:

```
. Processing [id 9504] child [id 231953]
. . Generating subtree clusters for [id 231953(species)]
. . . . [0 sqs] -- too few sequences, cannot make clusters
```

 * [ ] You might want a little "Don't Panic" comment saying that this is OK (it
 is, right?). 

Next we explore the results (still following the vignette):

```R
all_clustrs <- read_phylota(wd)
cids <- all_clustrs@cids
n_taxa <- get_ntaxa(phylota=all_clustrs, cid=cids)
```

I really liked the care that was put into making informative print statements.

 * [ ] The `show` and `print` generics are a little inconsistent. I recommend
 stepping through all your classes and ensuring that explicit calls to `print`,
 `show`, and the other generics are doing exactly what you want. For example:

```R
all_clustrs@sqs
# you probably want `cat` not `print` in `all_classes.R:301`
print(all_clustrs@sqs)
show(all_clustrs@sqs)
```

 * [ ] It might be worthwhile to print a one line description of each slot when
 printing an object. This would let the user interact with the objects without
 having to refer back to the docs. I'm not sure about this, though, since it
 might just be too cluttered.

Moving on:

``` R
keep <- cids[n_taxa > 6]
selected <- drop_cls(phylota=all_clustrs, cid=keep)
smmry <- summary(selected)
```

`summary(selected)` makes a table with a row for each table. I really like this
use of `summary`, good information, but two points:

 * [ ] You should document somewhere what MAD means. Also Type and Seed. 

 * [ ] In finding the most common words in the description and gene name, you
 might want to filter out common words. You might want to look into the `tm`
 library.

Continuing on to the last part of the vignette:

```R
# choose best sequence per species
reduced <- drop_by_rank(phylota=selected, rnk='species', n=1)
# get txids at the species level for each sequence
txids <- get_txids(phylota=reduced, cid=cid, rnk='species')
# look up name for txids
scientific_names <- get_tx_slot(phylota=reduced, txid=txids, slt_nm='scnm')
# clean the names
scientific_names <- gsub('\\.', '', scientific_names)
scientific_names <- gsub('\\s+', '_', scientific_names)
# look up sequence IDs for our chosen cluster
sids <- reduced@cls[[cid]]@sids
# write out
write_sqs(phylota=reduced, sid=sids, outfile='cytb.fasta',
          sq_nm=scientific_names)
```

 * [ ] I think `phylotaR` would benefit from reusing some existing libraries.
 `taxize` and `taxizedb` offer a lot of functions for mapping between IDs and
 scientific names. Also, for writing and reading FASTA sequencees, you might
 consider using the bioconductor Biostrings library.

 * [ ] FASTA files are normally formatted with on spaces between entries and with
 the sequences wrapped (usually at 60 or 80 characters). Again, you might want
 to use the Biostrings library.

**Comment by naupaka** [comment]
I am still working on my review. Should have it done in the next day or two.

**Comment by sckott** [comment]
thx for the update @naupaka

**Comment by sckott** [comment]
@naupaka Can you please get your review in soon, thanks

**Comment by naupaka** [comment]
I'm just about done with my review. Have had to let it run overnight a couple times and hit some snags. I'm hoping to have the review done by tomorrow late morning if all goes well on this current query.

**Comment by sckott** [comment]
thanks @naupaka

**Comment by naupaka** [comment]
## Package Review

- [X] As the reviewer I confirm that there are no conflicts of interest for me to review this work.


### Overall

I think this is a great and super useful package that I think is a good fit with ROpenSci. I am excited for this package to get finished up; I am already thinking up half a dozen ways to use it in my own research. All of that said, I had some problems getting it to run on my machine (even though it built and installed fine), and the documentation and examples are still being worked on. I'd be happy to take another look once it's further along, and like I said, I think this will be a really useful package for a whole lot of people, so I am grateful to the authors for putting it together. I agree with @arendsee's comments about considering more use of existing packages, especially `taxize` and `Biostrings` to simplify the code and reduce duplication of effort.

#### Documentation

The package includes all the following forms of documentation:

- [X] **A statement of need** clearly stating problems the software is designed to solve and its target audience in README

   There is a link to a website detailing this info, but it is not covered in the README in the repo.

- [X] **Installation instructions:** for the development version of package and any non-standard dependencies in README

   Installation was straighforward, despite the external dependency.

- [ ] **Vignette(s)** demonstrating major functionality that runs successfully locally

  There is a nice introductory vignette, but I had some trouble running through the examples (details below).

- [ ] **Function Documentation:** for all exported functions in R help

   The Description and Examples are still missing on a number of functions, e.g. `runTaxize()`.

- [ ] **Examples** for all exported functions in R Help that run successfully locally

   A good number of functions are still lacking examples.

- [ ] **Community guidelines** including contribution guidelines in the README or CONTRIBUTING, and DESCRIPTION with `URL`, `BugReports` and `Maintainer` (which may be autogenerated via `Authors@R`).

   No info on contributing yet, either in README or in a CONTRIBUTING file.

#### Functionality

- [X] **Installation:** Installation succeeds as documented.
- [ ] **Functionality:** Any functional claims of the software been confirmed.

   I cannot run through the pipeline suggessfully - I've tried with four different Taxa IDs and it fails each time on the `makeblastdb` step.

- [ ] **Performance:** Any performance claims of the software been confirmed.

   No claims made, but I think it would be great to include some ballpark info on how long a given run is going to take. If you don't `tail -f` the log file, it's hard to tell what's going on. The output in the R environment is very sparse.

- [ ] Automated tests: Unit tests cover essential functions of the package
   and a reasonable range of inputs and conditions. All tests pass on the local machine.

   Accourding to `covr()`, about half of the code is covered by tests.

- [ ] Packaging guidelines: The package conforms to the rOpenSci packaging guidelines

   Getting there!

#### Final approval (post-review)

- [ ] **The author has responded to my review and made changes to my satisfaction. I recommend approving this package.**

Estimated hours spent reviewing: 8 (not including overnight runs through the pipeline)

---

### Review Comments

- Does the code comply with general principles in the Mozilla reviewing guide?
   - I think so.
- Does the package comply with the ROpenSci packaging guide?
   - Getting there!
- Are there improvements that could be made to the code style?
   - I know it would be a good amount of work, but I think the function names could be standardized with a prefix, since some of them are very common terms and there could be confusion if they mask other functions of the same name. I also think the naming style in general is very condensed, and it wouldn't hurt to expand the names of functions and other objects a little in the codebase to aid in readability and facilitate contributions from others.
- Is there code duplication in the package that should be reduced?
   - Not as far as I can tell.
- Are there user interface improvements that could be made?
   - I think there could be a bit more in terms of output for the user when long-running functions are working. The log file on disk is great, and the disk file/folder organization is very neat and well-organized, but it's hard to tell what's going on in R. Since running the standard pipeline can take hours to tens of hours (in my experience), it would be nice to know all is still progressing.
- Are there performance improvements that could be made?
   - Don't think so. I imagine the rate-limiting step is the speed at which you can hit the NCBI API.
- Is the documentation (installation instructions/vignettes/examples/demos) clear and sufficient?
   - They are still in progress, and could use a bit more filling out (but I see this is still a work in progress).

We encourage you to use automated tools to facilitate your reviewing. These include:

- Running devtools::check() and devtools::test() on the package to find any errors that may be missed on the author's system.
   - No errors with either of these, but `R CMD check` returns a number of WARNINGS and NOTES.
- Using the goodpractice package (goodpractice::gp()) to identify likely sources of errors and style issues.
   - Already pasted above by @sckott
- Using devtools::spell_check() to find spelling errors in DESCRIPTION and manual files, and RStudio's spellchecker to find spelling errors in vignettes and README files.
   - Since these are still in progress, I imagine it would be more useful to check later.


#### Other Comments

```r
> devtools::session_info()

Session info -----------------------------------------------------------------------------------------------------------
 setting  value
 version  R version 3.4.4 (2018-03-15)
 system   x86_64, darwin17.3.0
 ui       RStudio (1.2.472)
 language (EN)
 collate  en_US.UTF-8
 tz       America/Los_Angeles
 date     2018-03-26

Packages ---------------------------------------------------------------------------------------------------------------
 package   * version date       source
 backports   1.1.2   2017-12-13 CRAN (R 3.4.3)
 base      * 3.4.4   2018-03-15 local
 base64enc   0.1-3   2015-07-28 CRAN (R 3.4.0)
 compiler    3.4.4   2018-03-15 local
 datasets  * 3.4.4   2018-03-15 local
 devtools    1.13.5  2018-02-18 CRAN (R 3.4.3)
 digest      0.6.15  2018-01-28 CRAN (R 3.4.3)
 evaluate    0.10.1  2017-06-24 CRAN (R 3.4.2)
 graphics  * 3.4.4   2018-03-15 local
 grDevices * 3.4.4   2018-03-15 local
 htmltools   0.3.6   2017-04-28 CRAN (R 3.4.0)
 jsonlite    1.5     2017-06-01 CRAN (R 3.4.0)
 knitr       1.20    2018-02-20 CRAN (R 3.4.3)
 magrittr    1.5     2014-11-22 CRAN (R 3.4.0)
 memoise     1.1.0   2017-04-21 CRAN (R 3.4.0)
 methods   * 3.4.4   2018-03-15 local
 phylotaR  * 0.1     2018-03-26 Github (DomBennett/phylotaR@2541b03)
 Rcpp        0.12.16 2018-03-13 CRAN (R 3.4.3)
 rmarkdown   1.9     2018-03-01 CRAN (R 3.4.3)
 rprojroot   1.3-2   2018-01-03 CRAN (R 3.4.3)
 stats     * 3.4.4   2018-03-15 local
 stringi     1.1.7   2018-03-12 cran (@1.1.7)
 stringr     1.3.0   2018-02-19 CRAN (R 3.4.3)
 sys         1.5     2017-10-10 cran (@1.5)
 tools       3.4.4   2018-03-15 local
 treeman     1.1.1   2017-06-27 cran (@1.1.1)
 utils     * 3.4.4   2018-03-15 local
 withr       2.1.2   2018-03-15 CRAN (R 3.4.3)
 yaml        2.1.18  2018-03-08 CRAN (R 3.4.3)

> sessionInfo()

R version 3.4.4 (2018-03-15)
Platform: x86_64-apple-darwin17.3.0 (64-bit)
Running under: macOS High Sierra 10.13.3

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libLAPACK.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] phylotaR_0.1

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.16    treeman_1.1.1   digest_0.6.15   withr_2.1.2     rprojroot_1.3-2 jsonlite_1.5    sys_1.5
 [8] backports_1.1.2 magrittr_1.5    evaluate_0.10.1 stringi_1.1.7   rstudioapi_0.7  rmarkdown_1.9   devtools_1.13.5
[15] tools_3.4.4     stringr_1.3.0   yaml_2.1.18     compiler_3.4.4  base64enc_0.1-3 memoise_1.1.0   htmltools_0.3.6
[22] knitr_1.20
```

#### Trying to run workflow pipeline on various taxa, encountering troubles

```r
> devtools::install_github(repo='DomBennett/phylotaR', build_vignettes=TRUE)
# https://github.com/DomBennett/phylotaR/commit/2541b030158183b8b1d3c1f44c0ea26e96ab11f3
library("phylotaR")
```

```r
wd <- "~/Desktop/phylota_review/primates"
ncbi_dr <- "/usr/local/bin/"
txid <- 9443 #primates
setUp(wd=wd, txid=txid, ncbi_dr=ncbi_dr)
run(wd=wd)
```

Hung when trying to download chimpanzees, no progress in log file for 20+ minutes.

> . . Working on child [id 9598]
> . . + direct ...

Trying again with a smaller clade: Mycosphaerella (fungal genus) -- 41254

```r
wd <- "~/Desktop/phylota_review/fungi"
ncbi_dr <- "."
txid <- 41254 #Mycosphaerella
setUp(wd=wd, txid=txid, ncbi_dr=ncbi_dr)
```

`ncbi_dr <- "."` did not work as recommended in the vignette despite the programs being in my `PATH`.

```
Error in setUpNcbiTools(d = ncbi_dr, v = v, wd = wd) : Unable to find correct versions of NCBI BLAST+ tools
3. stop(msg)
2. setUpNcbiTools(d = ncbi_dr, v = v, wd = wd)
1. setUp(wd = wd, txid = txid, ncbi_dr = ncbi_dr)
```

Works fine with absolute path to directory.

```r
wd <- "~/Desktop/phylota_review/fungi"
ncbi_dr <- "/usr/local/bin"
txid <- 41254 #Mycosphaerella (large genus)
setUp(wd=wd, txid=txid, ncbi_dr=ncbi_dr)
run(wd=wd)
```

I kept getting the following error, even on different initial query taxa IDs.

```
Error in runStgs(wd = wd, frm = 1, to = nstages, stgs_msg = stgs_msg) : Unexpected Error in error(ps = ps, paste0("makeblastdb failed to run. Check BLAST log files.")) : Error: makeblastdb failed to run. Check BLAST log files. Occurred [2018-03-25 20:45:13] Contact package maintainer for help.
3. stop(msg)
2. runStgs(wd = wd, frm = 1, to = nstages, stgs_msg = stgs_msg)
1. run(wd = wd)
```

```
---------------------------------------------
Starting stage CLUSTER: [2018-03-25 20:45:12]
---------------------------------------------
Working on [id 41254]
. Generating subtree clusters for [id 41254(genus)]
. Generating direct clusters for [id 41254(genus)]
. . [0 sqs]
. . . Too few sequences, cannot make clusters
. BLASTing [1079 sqs] ....
. . Running makeblastdb
Error: makeblastdb failed to run. Check BLAST log files.
Unexpected Error in error(ps = ps, paste0("makeblastdb failed to run. Check BLAST log files.")) :
  Error: makeblastdb failed to run. Check BLAST log files.


Occurred [2018-03-25 20:45:13]
Contact package maintainer for help.
```

Didn't work because not enough sequence diversity. Trying again with a higher level group.

```r
wd <- "~/Desktop/phylota_review/capnodiales"
ncbi_dr <- "/usr/local/bin"
txid <- 134362 # Capnodiales (fungal family)
setUp(wd=wd, txid=txid, ncbi_dr=ncbi_dr)
run(wd=wd)
```

```
... Taxise
... Download
... Cluster
Error in runStgs(wd = wd, frm = 1, to = nstages, stgs_msg = stgs_msg) :
  Unexpected Error in error(ps = ps, paste0("makeblastdb failed to run. Check BLAST log files.")) :
  Error: makeblastdb failed to run. Check BLAST log files.


Occurred [2018-03-25 23:30:10]
Contact package maintainer for help.
```

Here's the output from the blast log file.

```
# blast log file
$ cat blast/taxon-1047167-typ-subtree-db.log

BLAST options error: File ~/Desktop/phylota_review/capnodiales/blast/taxon-1047167-typ-subtree-db.fa does not exist
w/capnodiales/blast/taxon-1047167-typ-subtree-db.fa
New DB title:  ~/Desktop/phylota_review/capnodiales/blast/taxon-1047167-typ-subtree-db.fa
Sequence type: Nucleotide
Keep MBits: T
Maximum file size: 1000000000B
```

Some thoughts on this process:

* It's unclear what makes a suitable clade (during the search process).
* There is helpful info on how many out of how many parental clades have been downloaded, but since it's unclear how many children for each there might be and how many sequences in each of those, it is hard to tell how long it will all take. Is there any way to estimate and report this before starting?
* For (e.g.) genera with many species in them, it is unclear why the dots in the logfile keep increasing. If all the species are sister to one another, shouldn't they not be indented under one another?

```r
restart(wd)
```

Does not restart wthout restarting the R session.

#### Trying to run demo in aotus.R:

Also using a absolute instead of a relative path for the working directory here, in case that is screwing up `makeblastdb`, but it still didn't work, as before.

```
> # RUN PIPELINE
> txid <- 9504
> setUp(wd="/Users/naupaka/Desktop/phylota_review/aotus", txid=txid, ncbi_dr=ncbi_dr, v=TRUE)
-----------------------------------------------
phylotaR: Implementation of PhyLoTa in R [v0.1]
-----------------------------------------------
Checking for valid NCBI BLAST+ Tools ...
Found: [/usr/local/bin/makeblastdb]
Found: [/usr/local/bin/blastn]
Setting up pipeline with the following parameters:
. blstn      [/usr/local/bin/blastn]
. btchsz     [300]
. date       [2018-03-26]
. mdlthrs    [3000]
. mkblstdb   [/usr/local/bin/makeblastdb]
. mncvrg     [51]
. mnsql      [250]
. mxevl      [1e-10]
. mxnds      [1e+05]
. mxrtry     [100]
. mxsql      [2000]
. mxsqs      [50000]
. ncps       [1]
. txid       [9504]
. v          [TRUE]
. wd         [/Users/naupaka/Desktop/phylota_review/aotus]
-----------------------------------------------
> run(wd=wd)
... Taxise
... Download
... Cluster
Error in runStgs(wd = wd, frm = 1, to = nstages, stgs_msg = stgs_msg) :
  Unexpected Error in error(ps = ps, paste0("makeblastdb failed to run. Check BLAST log files.")) :
  Error: makeblastdb failed to run. Check BLAST log files.


Occurred [2018-03-26 06:37:35]
Contact package maintainer for help.
```

#### Output of R CMD check

```sh
$ R CMD check .
* using log directory ‘/Users/naupaka/Desktop/phylota_review/phylotaR/..Rcheck’
* using R version 3.4.4 (2018-03-15)
* using platform: x86_64-apple-darwin17.3.0 (64-bit)
* using session charset: UTF-8
* checking for file ‘./DESCRIPTION’ ... OK
* checking extension type ... Package
* this is package ‘phylotaR’ version ‘0.1’
* checking package namespace information ... OK
* checking package dependencies ... OK
* checking if this is a source package ... OK
* checking if there is a namespace ... OK
* checking for executable files ... OK
* checking for hidden files and directories ... NOTE
Found the following hidden files and directories:
  .gitignore
  .travis.yml
  ..Rcheck
  .git
These were most likely included in error. See section ‘Package
structure’ in the ‘Writing R Extensions’ manual.
* checking for portable file names ... OK
* checking for sufficient/correct file permissions ... OK
* checking whether package ‘phylotaR’ can be installed ... OK
* checking installed package size ... OK
* checking package directory ... OK
* checking DESCRIPTION meta-information ... NOTE
Checking should be performed on sources prepared by ‘R CMD build’.
* checking top-level files ... OK
* checking for left-over files ... OK
* checking index information ... OK
* checking package subdirectories ... WARNING
Found the following directory with the name of a check directory:
  ./..Rcheck
Most likely, these were included erroneously.
* checking R files for non-ASCII characters ... OK
* checking R files for syntax errors ... OK
* checking whether the package can be loaded ... OK
* checking whether the package can be loaded with stated dependencies ... OK
* checking whether the package can be unloaded cleanly ... OK
* checking whether the namespace can be loaded with stated dependencies ... OK
* checking whether the namespace can be unloaded cleanly ... OK
* checking loading without being on the library search path ... OK
* checking dependencies in R code ... OK
* checking S3 generic/method consistency ... OK
* checking replacement functions ... OK
* checking foreign function calls ... OK
* checking R code for possible problems ... OK
* checking Rd files ... OK
* checking Rd metadata ... OK
* checking Rd cross-references ... OK
* checking for missing documentation entries ... OK
* checking for code/documentation mismatches ... OK
* checking Rd \usage sections ... OK
* checking Rd contents ... OK
* checking for unstated dependencies in examples ... OK
* checking contents of ‘data’ directory ... OK
* checking data for non-ASCII characters ... OK
* checking data for ASCII and uncompressed saves ... OK
* checking files in ‘vignettes’ ... WARNING
Files in the 'vignettes' directory but no files in 'inst/doc':
  ‘phylotaR-advanced.Rmd’, ‘phylotaR-advanced.html’, ‘phylotaR.R’,
    ‘phylotaR.Rmd’, ‘phylotaR.html’
* checking examples ... OK
* checking for unstated dependencies in ‘tests’ ... OK
* checking tests ...
  Running ‘test-all.R’
 OK
* checking for unstated dependencies in vignettes ... OK
* checking package vignettes in ‘inst/doc’ ... WARNING
Package vignettes without corresponding PDF/HTML:
   ‘phylotaR-advanced.Rmd’
   ‘phylotaR.Rmd’

* checking running R code from vignettes ...
   ‘phylotaR-advanced.Rmd’ using ‘UTF-8’ ... OK
   ‘phylotaR.Rmd’ using ‘UTF-8’ ... OK
 NONE
* checking re-building of vignette outputs ...Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
Warning in file.copy(pkgdir, vd2, recursive = TRUE) : too deep nesting
 OK
* checking PDF version of manual ... OK
* DONE

Status: 3 WARNINGs, 2 NOTEs
See
  ‘/Users/naupaka/Desktop/phylota_review/phylotaR/..Rcheck/00check.log’
for details.
```

**Comment by sckott** [comment]
thanks very much for your review @naupaka !!!

**Comment by DomBennett** [comment]
Thanks very much @arendsee and @naupaka for your detailed reviews!

I feel I may have already addressed some of the run issues raised by @naupaka, as I encountered some of them over the past few weeks myself. The hanging is particularly annoying and seems to be due to Entrez, I've added a timeout so that it will retry after one hour. Also, I wonder if the BLAST errors might be due to using Windows, I've not tested the pipeline at all on that OS. Lots of things to be getting on with!

In terms of next steps @sckott , do I just update the package and report what I did here?

**Comment by sckott** [comment]
@DomBennett  So we don't have a rule on how exactly to respond to reviewers. This is how it's best approached though I think:

Open issues in your repo that address issues raised here - Then when you reply to reviewers you can link to those - That's what I do when I bring my own pkgs through review, and I use a label or wording in the issue title to indicate its from the review, e.g. https://github.com/ropensci/charlatan/issues?q=is%3Aissue+is%3Aopen+label%3Areview What's good about opening issues is you can reference those in your commits so that everything is all linked up. It's likely that you'll want to ask for additional feedback from reviewers - so being able to link to an issue that is linked to the code itself is a nice approach

**Comment by DomBennett** [comment]
Hi @sckott,

I've copied your format and created [issues](https://github.com/AntonelliLab/phylotaR/issues). Also, FYI we have now submitted a manuscript describing phylotaR to MDPI Life, preprint available [here](https://www.preprints.org/manuscript/201804.0047/v1).

Is there a deadline for responding to the reviewers? I was thinking of taking a break and coming back to phylotaR with a little more distance.

**Comment by sckott** [comment]
Not really a deadline. If you'll put it off for a while, we have a [holding label](https://github.com/ropensci/onboarding/issues?q=is%3Aissue+is%3Aopen+label%3Aholding) that we'll use, and then we'll know to not ask you about progress until you're ready to come back to it

**Comment by DomBennett** [comment]
Hi @sckott, @arendsee and @naupaka,

I’ve made the following major changes as per your comments and reviews:

- **Restyled**:  All functions and variables are snake case with the exception of the S4 classes that remain camelcase. I’ve also updated the names to make things a little more explicit (e.g. ClstrRec rather than ClRec).
- **Tested**: I now have [89% coverage](https://coveralls.io/github/AntonelliLab/phylotaR?branch=master)
- **Examples**: All ‘public’, exported functions and data have examples.
- **Pkgdown**: I have created a [website](https://antonellilab.github.io/phylotaR/).
- **Documentation**: all functions have return parameters, I have checked spelling with spell_check.
- **Timings**: The demonstration phylotaR runs now have the timings of each stage on the README.md
- **Contributing**: There is now a CONTRIBUTING.md
- **Logging**: There is now more salient information printed to screen, blast versions and session info are also recorded in a user’s wd.
- **Windows**: The pipeline has been run on a windows machine
- **Makeblastdb error**: I think this was due to a ‘~’ in @naupaka’s filepath. I now halt the pipeline if a user’s does this.

Things I have not done:

- **Additional packages**: I have not incorporated taxize, taxizeDB, tm or Biostrings into the package. In the case of the latter, I feel it best to avoid Bioconductor packages as they are a little more tricky to install and I would like to keep phylotaR light, even though it requires BLAST+. In the case of the others, I see no immediate reason to make use of these additional packages while I currently have no need for their additional capabilities. I have fixed the 80 column width issue of the user generated .fasta output. In the future, as new features are requested (e.g. alternative taxonomies), then they should be incorporated. Right now though, I feel I should follow the doctrine of if it ain’t broke, why fix it?
- **Advanced vignette**: I have not added this as a lot of the additional information can be found in the recently [published article](https://doi.org/10.3390/life8020020) on the pipeline. In the future I can imagine a new vignette containing more detailed information on new features with more complex options. This is likely to occur since I have been contacted a fair number of times with requests for new options and features.

If any of you feel that these last two points should be acted upon in some form, then I will be happy to look at these again.

Many thanks once more for taking the time to review the package!

Best wishes,
Dom

**Comment by DomBennett** [comment]
Oh also... I should add that I cannot rename the first stage 'taxise' as it is already published (I interpret taxise to just mean grab taxonomic information or place in a taxonomic context.). To compromise I have updated the stage diagram in the README to be a more explicit.

**Comment by naupaka** [comment]
Sounds great @DomBennett!

@sckott shall we have another go through of it with the new version? What's the ROpenSci SOP on review after revisions?

**Comment by sckott** [comment]
thanks @DomBennett ! 

@naupaka @arendsee 

If you're completely happy with changes to the package and responses from the maintainer then just say so. If not, please do comment here with any additional comments/suggestions/etc.

**Comment by arendsee** [comment]
@DomBennett @sckott Looks good to me.

I would recommend *eventually* refactoring with `taxize` et al, but I wouldn't consider it a necessity.

Good job!

**Comment by sckott** [comment]
@naupaka do you want to make further comments, or are you happy?

**Comment by naupaka** [comment]
I'd like to put the new version through its paces and haven't had the time to sit down and do it. Will get back to you in the next couple days if that's ok?

**Comment by sckott** [comment]
@naupaka  sounds good

**Comment by naupaka** [comment]
@DomBennett Just ran through a couple testing workflows and everything works smoothly for me! Thanks for this very useful tool.

**Comment by naupaka** [comment]
@sckott I'm happy to give this my stamp of approval 👍

**Comment by sckott** [comment]
Great, thanks again @arendsee and @naupaka for your reviews!

**Comment by sckott** [comment]
@DomBennett we're nearly there. just a few comments:

On running the example from the vignette, I get an error: 
 
```r
---------------------------------------------------
Running pipeline on [unix] at [2018-07-09 10:22:31]
---------------------------------------------------
Running stages: taxise, download, cluster, cluster2
--------------------------------------------
Starting stage TAXISE: [2018-07-09 10:22:31]
--------------------------------------------
Searching taxonomic IDs ...
Downloading taxonomic records ...
. [1-21]
Generating taxonomic dictionary ...
---------------------------------------------
Completed stage TAXISE: [2018-07-09 10:22:33]
---------------------------------------------
----------------------------------------------
Starting stage DOWNLOAD: [2018-07-09 10:22:33]
----------------------------------------------
Identifying suitable clades ...
Identified [1] suitable clades.
Downloading hierarchically ...
Working on parent [id 9504]: [1/1] ...
. + whole subtree ...
. . Getting [2700 sqs] ...
. . . [1-300]
Unexpected Error : NCBI is limiting the size of your request. Consider reducing btchsz.


Occurred [2018-07-09 10:22:58]
Contact package maintainer for help.
Error in stages_run(wd = wd, frm = 1, to = nstages, stgs_msg = stgs_msg) :
  Unexpected Error : NCBI is limiting the size of your request. Consider reducing btchsz.


Occurred [2018-07-09 10:22:58]
Contact package maintainer for help.
```

It looks like I need to fiddle with `btchsz` but I'm not sure where to do that. Would it make sense to say what function has that parameter? It looks like it's the function `parameters()`? 

running goodpractice again: i'd encourage you to avoid `sapply` where possible and avoid 1:length(...), 1:nrow(...), 1:ncol(...), 1:NROW(...) and 1:NCOL(...) type expressions where possible,  but not there aren't required before approval.

**Comment by sckott** [comment]
Approved!  Thanks again for your submission @DomBennett 

* Please transfer the package  to `ropensci`- I've invited you to a team within our ropensci organization. After you accept you should be able to move it. You'll be made admin once you do.
* Make sure to 
    * add rOpenSci footer to README
`[![ropensci_footer](https://ropensci.org/public_images/ropensci_footer.png)](https://ropensci.org)`
    * Change any needed links, such those for CI badges
    * Travis CI should update to the new location automatically - you may have to update other CI systems manually
* In a new `.github` folder in the root of the pkg, add a `CONTRIBUTING.md` file, an issue template file, and a PR template file, see https://github.com/ropensci/dotgithubfiles/ for examples, you can copy them directly and edit as you wish
* We're starting to roll out software metadata files to all ropensci pkgs via the Codemeta initiative, see  https://github.com/ropensci/codemetar/#codemetar  for how to include it in your pkg, after installing the pkg - should be easy as running `codemetar::write_codemeta()` in the root of your pkg

---

* Are you interested in doing a blog post for our blog https://ropensci.org/blog/ ?  either a short-form intro to it (https://ropensci.org/technotes/) or long-form post with more narrative about its development (https://ropensci.org/blog/). If so, we'll have our community manager @stefaniebutland get in touch with you on that

**Comment by DomBennett** [comment]
Done!
https://github.com/ropensci/phylotaR

**Comment by sckott** [comment]
@DomBennett and the question about a blog post?

**Comment by stefaniebutland** [comment]
@DomBennett this link will give you many examples of blog posts by authors of onboarded packages so you can get an idea of the style and length: https://ropensci.org/tags/review/. Technotes are here: https://ropensci.org/technotes/.

Here are some technical and editorial guidelines for contributing a blog post: https://github.com/ropensci/roweb2#contributing-a-blog-post. 

Happy to answer any questions.
